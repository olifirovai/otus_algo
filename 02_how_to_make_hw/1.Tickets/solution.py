import datetime as dt
import os

from termcolor import colored

TEST_PATH = f'{os.getcwd()}\\tests'


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    for dirpath, dirnames, filenames in os.walk(TEST_PATH):
        for filename in filenames:
            if filename.endswith('in'):
                test_in.append(filename)
            elif filename.endswith('out'):
                test_out.append(filename)

    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    for one_case in range(len(test_in)):
        answer = int(open(f'{TEST_PATH}\\{test_out[one_case]}', 'r').read())
        data = int(
            open(f'{TEST_PATH}\\{test_in[one_case]}', 'r').read().strip())

        start_time = dt.datetime.now()

        # here the main function
        result = sum([x ** 2 for x in lucky_tickets(data)])

        func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
        output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                      f'finished with {result == answer} result'
        if result == answer:
            print(colored(output_text, 'green'))

        else:
            print(colored(output_text, 'red'))
        print('----------------------------------------------------')


def lucky_tickets(n: int, line=[1]):
    new_line = []
    if n > 1:
        line = lucky_tickets(n - 1, line)

    for i in range(9 * n + 1):
        new = 0
        for j in range(max(0, len(line) - 1 - i),
                       min(len(line), 9 * n + 1 - i)):
            new += line[j]
        new_line.append(new)

    # return the last list which is consisted of sums for the half of the
    # whole ticket
    # ex. for n = 2: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
    return new_line


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
