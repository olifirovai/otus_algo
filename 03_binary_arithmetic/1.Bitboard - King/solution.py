import datetime as dt
import os
from typing import Tuple

from termcolor import colored

TEST_PATH = f'{os.getcwd()}\\tests'


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    for dirpath, dirnames, filenames in os.walk(TEST_PATH):
        for filename in filenames:
            if filename.endswith('in'):
                test_in.append(filename)
            elif filename.endswith('out'):
                test_out.append(filename)

    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    for one_case in range(len(test_in)):
        with open(f'{TEST_PATH}\\{test_out[one_case]}', "r") as test_file:
            answer = [int(line.rstrip('\n')) for line in test_file]
        data = int(
            open(f'{TEST_PATH}\\{test_in[one_case]}', "r").read().strip())

        start_time = dt.datetime.now()

        # here the main function
        result = list(king_turn(data))
        func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
        output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                      f'finished with {result == answer} result'
        if result == answer:
            print(colored(output_text, 'green'))

        else:
            print(colored(output_text, 'red'))
        print('----------------------------------------------------')


def king_turn(position: int) -> Tuple[int, int]:
    king = 1 << position
    no_a = 0xfefefefefefefefe
    no_h = 0x7f7f7f7f7f7f7f7f
    whole_board = 0xffffffffffffffff

    king_a = king & no_a
    king_h = king & no_h
    moves = ((king_a << 7) | (king << 8) | (king_h << 9) |
             (king_a >> 1) | (king_h << 1) |
             (king_a >> 9) | (king >> 8) | (king_h >> 7)) & whole_board

    return count_bits(moves), moves


def count_bits(moves):
    count = 0
    while moves > 0:
        count += 1
        moves &= moves - 1
    return count


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
