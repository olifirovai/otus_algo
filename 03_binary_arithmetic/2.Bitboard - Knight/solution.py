import datetime as dt
import os
from typing import Tuple

from termcolor import colored

TEST_PATH = f'{os.getcwd()}\\tests'


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    for dirpath, dirnames, filenames in os.walk(TEST_PATH):
        for filename in filenames:
            if filename.endswith('in'):
                test_in.append(filename)
            elif filename.endswith('out'):
                test_out.append(filename)

    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    for one_case in range(len(test_in)):
        with open(f'{TEST_PATH}\\{test_out[one_case]}', "r") as test_file:
            answer = [int(line.rstrip('\n')) for line in test_file]
        data = int(
            open(f'{TEST_PATH}\\{test_in[one_case]}', "r").read().strip())

        start_time = dt.datetime.now()

        # here the main function
        result = list(knight_turn(data))

        func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
        output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                      f'finished with {result == answer} result'
        if result == answer:
            print(colored(output_text, 'green'))

        else:
            print(f'data = {data}')
            print(f'answer = {answer}')
            print(f'my result = {result}')
            print(colored(output_text, 'red'))
        print('----------------------------------------------------')


def knight_turn(position: int) -> Tuple[int, int]:
    knight = 1 << position
    no_a_b = 0xfcfcfcfcfcfcfcfc
    no_b = 0xfdfdfdfdfdfdfdfd
    no_h = 0x7f7f7f7f7f7f7f7f
    no_g = 0xbfbfbfbfbfbfbfbf

    knight_a = knight & no_g & no_h
    knight_b = knight & no_h
    knight_g = knight & no_a
    knight_h = knight & no_a & no_b

    moves = ((knight_b << 17) | (knight_g << 15) |
(knight_a << 10) |                              (knight_h << 6) |
(knight_a >> 10) |                              (knight_h >> 6) |
             (knight_b >> 17) | (knight_g >> 15))

    return count_bits(moves), moves


def count_bits(moves):
    count = 0
    while moves > 0:
        count += 1
        moves &= moves - 1
    return count


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
