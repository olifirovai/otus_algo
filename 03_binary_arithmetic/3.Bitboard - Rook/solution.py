import datetime as dt
import os

from termcolor import colored


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    for dirpath, dirnames, filenames in os.walk(os.getcwd()):
        for filename in filenames:
            if filename.endswith('in'):
                test_in.append(filename)
            elif filename.endswith('out'):
                test_out.append(filename)

    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    for one_case in range(len(test_in)):
        answer = int(open(test_out[one_case], "r").read())
        data = int(open(test_in[one_case], "r").read().strip())

        start_time = dt.datetime.now()

        # here the main function
        result = rook_turn(data)

        func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
        output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                      f'finished with {result == answer} result'
        if result == answer:
            print(colored(output_text, 'green'))

        else:
            print(colored(output_text, 'red'))
        print('----------------------------------------------------')


def rook_turn(n: int) -> int:
    pass


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
