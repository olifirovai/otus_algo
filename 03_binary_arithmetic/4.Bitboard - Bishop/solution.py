import datetime as dt
import os
from typing import Tuple

from termcolor import colored

TEST_PATH = f'{os.getcwd()}\\tests'


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    for dirpath, dirnames, filenames in os.walk(TEST_PATH):
        for filename in filenames:
            if filename.endswith('in'):
                test_in.append(filename)
            elif filename.endswith('out'):
                test_out.append(filename)

    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    for one_case in range(len(test_in)):
        with open(f'{TEST_PATH}\\{test_out[one_case]}', "r") as test_file:
            answer = [int(line.rstrip('\n')) for line in test_file]
        data = int(
            open(f'{TEST_PATH}\\{test_in[one_case]}', "r").read().strip())

        start_time = dt.datetime.now()

        # here the main function
        result = list(bishop_turn(data))

        func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
        output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                      f'finished with {result == answer} result'
        if result == answer:
            print(colored(output_text, 'green'))

        else:
            print(colored(output_text, 'red'))
        print('----------------------------------------------------')


def bishop_turn(position: int) -> Tuple[int, int]:
    bishop = 1 << position
    no_a = 0xfefefefefefefefe
    no_h = 0x7f7f7f7f7f7f7f7f
    whole_board = 0xffffffffffffffff

    up_left_move = (bishop << 7) & no_h
    up_right_move = (bishop << 9) & no_a
    down_left_move = (bishop >> 9) & no_h
    down_right_move = (bishop >> 7) & no_a

    for _ in range(7):
        up_left_move = (up_left_move | up_left_move << 7) & no_h
        up_right_move = (up_right_move | up_right_move << 9) & no_a
        down_left_move = (down_left_move | down_left_move >> 9) & no_h
        down_right_move = (down_right_move | down_right_move >> 7) & no_a

    moves = (up_left_move | up_right_move | down_left_move | down_right_move) & whole_board

    return count_bits(moves), moves


def count_bits(moves):
    count = 0
    while moves > 0:
        count += 1
        moves &= moves - 1
    return count


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
