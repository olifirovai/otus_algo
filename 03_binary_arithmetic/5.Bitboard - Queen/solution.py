import datetime as dt
import os

from termcolor import colored

TEST_PATH = f'{os.getcwd()}\\tests'


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    for dirpath, dirnames, filenames in os.walk(TEST_PATH):
        for filename in filenames:
            if filename.endswith('in'):
                test_in.append(filename)
            elif filename.endswith('out'):
                test_out.append(filename)

    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    for one_case in range(len(test_in)):
        with open(f'{TEST_PATH}\\{test_out[one_case]}', "r") as test_file:
            answer = [int(line.rstrip('\n')) for line in test_file]
        data = int(
            open(f'{TEST_PATH}\\{test_in[one_case]}', "r").read().strip())

        start_time = dt.datetime.now()

        # here the main function
        result = list(queen_turn(data))

        func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
        output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                      f'finished with {result == answer} result'
        if result == answer:
            print(colored(output_text, 'green'))

        else:
            print(colored(output_text, 'red'))
        print('----------------------------------------------------')


def queen_turn(n: int) -> int:
    pass


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
