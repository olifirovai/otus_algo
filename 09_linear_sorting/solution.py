def quick_sort(data: list) -> list:
    if len(data) < 2:
        return data
    else:
        # Let's pick a center index of the array
        pivot = len(data) // 2
        less, center, greater = partition(data, pivot)
        return quick_sort(less) + center + quick_sort(greater)


def partition(array: list, pivot: int):
    less = []
    center = []
    greater = []
    for i in range(len(array)):
        if array[i] < array[pivot]:
            less.append(array[i])
        elif array[i] > array[pivot]:
            greater.append(array[i])
        else:
            center.append(array[i])

    return less, center, greater


def merge_sort(data: list) -> list:
    if len(data) == 1:
        return data
    left = merge_sort(data[:len(data) // 2])
    right = merge_sort(data[len(data) // 2:])

    l, r, k = 0, 0, 0

    while l < len(left) and r < len(right):
        if left[l] <= right[r]:
            data[k] = left[l]
            l += 1
        else:
            data[k] = right[r]
            r += 1
        k += 1
    while l < len(left):
        data[k] = left[l]
        l += 1
        k += 1

    while r < len(right):
        data[k] = right[r]
        r += 1
        k += 1

    return data


def mix_merge_quick_sort(data: list) -> list:
    if len(data) < 1024:
        return quick_sort(data)

    return merge_sort(data)


def counting_sort(data: list) -> list:
    return data


def radix_sort(data: list) -> list:
    max_number = max(data)

    digit_place = 1

    lists_list = []

    for _ in range(10):
        lists_list.append([])

    while max_number / digit_place >= 1:

        for number in data:
            list_index = (number // digit_place) % 10

            lists_list[list_index].append(number)

        data.clear()

        for single_list in lists_list:
            data.extend(single_list)

            single_list.clear()

        digit_place *= 10

    return data


def bucket_sort(data: list) -> list:
    return data
