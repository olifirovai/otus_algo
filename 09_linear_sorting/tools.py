import json
import os
import struct

LENGTH_LIST = {'0': 10, '1': 1000, '2': 100000, '3': 1000000, '4': 10000000,
               '5': 100000000}

TEST_PATH = f'{os.getcwd()}\\tests'

FUNC_DICT = {'merge_sort': 'Merge sort',
             'mix_merge_quick_sort': 'Merge & Quick sort',
             'counting_sort': 'Counting sort', 'radix_sort': 'Radix sort',
             'bucket_sort': 'Bucket sort'}


def create_dict_txt(test_in, func_dict):
    with open('result_dict.txt', 'w') as s:
        result_dict = {}
        for func in func_dict.values():
            result_dict[func] = {}
            for test in range(len(test_in)):
                test_num = f'{test}'
                result_dict[func][test_num] = 'Too much'
        s.write(json.dumps(result_dict))


def load_data_txt(one_case, func_dict, func, func_time):
    with open('result_dict.txt', 'r') as f:
        data = f.read()
        result_dict = json.loads(data)

    with open('result_dict.txt', 'w') as s:
        test_num = f'{one_case}'
        func_disc = func_dict[func]
        result_dict[func_disc][test_num] = func_time
        s.write(json.dumps(result_dict))


def print_table_readme(file):
    with open(file, 'r') as f:
        data = f.read()
        result_dict = json.loads(data)

    list_sort = list(result_dict.keys())
    header = '|Test â„–|Length Array|'
    header_line = '|---|---|'
    for one_sort in list_sort:
        header += f'{one_sort}|'
        header_line += '---|'
    print(header)
    print(header_line)

    for test in range(len((list(LENGTH_LIST.keys())))):
        line = f'|{test}|{LENGTH_LIST[str(test)]}'
        for sort in list_sort:
            time = result_dict[sort][str(test)]
            line += f'|{time} ms'
        print(f'{line}|')


def unpack_bin_file(data: bytes, length: int) -> list:
    number_list = list(struct.unpack(f'@{length}h', data))

    # make integers between 0 and 65536
    for i in range(len(number_list)):
        number_list[i] += 2 ** 15

    return number_list


def pack_bin_file(sorted_list, length: int):
    test_number = list(LENGTH_LIST.values()).index(length)
    new_bin_file = open(f'{TEST_PATH}\\test.{test_number}.out', 'wb')

    # make integers between -32768 and 32768
    for i in range(length):
        sorted_list[i] -= 2 ** 15

    bin_list = struct.pack(f'@{length}h', *sorted_list)
    new_bin_file.write(bin_list)
