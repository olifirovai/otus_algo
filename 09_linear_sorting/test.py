import datetime as dt

from natsort import os_sorted
from termcolor import colored

from solution import *  # noqa
from tools import *

AMOUNT_TESTS = 3


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []
    filenames = os_sorted(os.listdir(f'{TEST_PATH}'))
    for filename in filenames:
        if filename.endswith('in'):
            test_in.append(filename)
        elif filename.endswith('out'):
            test_out.append(filename)
    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    if 'result_dict.txt' not in os.listdir(os.getcwd()):
        create_dict_txt(test_in, FUNC_DICT)

    for func in FUNC_DICT.keys():
        # if func != 'radix_sort':
        #     continue
        print(
            colored(f'Tests with function {func}', 'magenta', attrs=['bold']))
        print('----------------------------------------------------')

        for one_case in range(AMOUNT_TESTS):
            length = LENGTH_LIST[f'{one_case}']
            # if length != 100000000:
            #     continue
            data = open(f'{TEST_PATH}\\{test_in[one_case]}', "rb").read()
            data = unpack_bin_file(data, length)
            answer = open(f'{TEST_PATH}\\{test_out[one_case]}', "rb").read()
            answer = unpack_bin_file(answer, length)
            start_time = dt.datetime.now()

            # here the main function
            result = eval(f'{func}({data})')

            func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
            load_data_txt(one_case, FUNC_DICT, func, func_time)
            output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                          f'finished with {result == answer} result'
            if result == answer:
                print(colored(output_text, 'green'))

            else:
                # print(f'data {data}')
                # print(f'my result {result}')
                # print(f'answer {answer}')
                print(colored(output_text, 'red'))
            print('----------------------------------------------------')


def main():
    # test_in, test_out = find_test_cases()
    # make_tests(test_in, test_out)
    print_table_readme('result_dict.txt', )


if __name__ == '__main__':
    main()
