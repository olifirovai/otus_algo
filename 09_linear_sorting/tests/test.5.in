version https://git-lfs.github.com/spec/v1
oid sha256:9b742d914871e346861efe3a974cb86ae5fef7b63090f9fac875741b6f25cb01
size 200000000
