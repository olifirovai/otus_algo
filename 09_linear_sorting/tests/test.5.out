version https://git-lfs.github.com/spec/v1
oid sha256:f18c099786ffbc09f1a6f7a36c3be65775d181315d5dfc39442a86d0aa6606f1
size 200000000
